File: scala-servers/EtherFlowClient.scala
// EtherFlowClient.scala
// EtherFlow HTTP Client for Scala — sttp + circe
//
// A strongly-typed, Future-based HTTP client for Scala 3 that mirrors the
// EtherFlow Java/Kotlin builder API. Uses sttp for transport and circe for JSON.
//
// Usage:
//   val client = EtherFlowClient.builder
//     .baseUrl("https://api.example.com")
//     .retry(3)
//     .build
//
//   val user: Future[User] = client.get[User]("/users/1")
//   val created: Future[User] = client.post[User, NewUser]("/users", newUser)

package io.etherflow.scala

import sttp.client4.*
import sttp.client4.circe.*
import sttp.model.*
import io.circe.*
import io.circe.generic.auto.*
import io.circe.syntax.*

import scala.concurrent.*
import scala.concurrent.duration.*
import scala.util.*

// ─────────────────────────────────────────────────────────────────────────────
// Error
// ─────────────────────────────────────────────────────────────────────────────

final case class EtherFlowError(
  message: String,
  statusCode: Option[Int] = None,
  responseBody: Option[String] = None,
  cause: Option[Throwable] = None
) extends Exception(message, cause.orNull)

// ─────────────────────────────────────────────────────────────────────────────
// Config
// ─────────────────────────────────────────────────────────────────────────────

final case class EtherFlowConfig(
  baseUrl: String                        = "",
  timeout: Duration                      = 10.seconds,
  maxRetries: Int                        = 3,
  retryDelay: Duration                   = 200.milliseconds,
  defaultHeaders: Map[String, String]    = Map.empty
)

// ─────────────────────────────────────────────────────────────────────────────
// Safe Result (mirrors EtherFlow Java's Mono<Result<T>>)
// ─────────────────────────────────────────────────────────────────────────────

enum EtherFlowResult[+T]:
  case Success(data: T)
  case Failure(error: EtherFlowError)

// ─────────────────────────────────────────────────────────────────────────────
// Client
// ─────────────────────────────────────────────────────────────────────────────

/** EtherFlow HTTP client for Scala 3. Mirrors the Java/Kotlin EtherFlowClient builder API. */
final class EtherFlowClient private (config: EtherFlowConfig)(using ExecutionContext):

  private val backend: SttpBackend[Future, Any] =
    HttpClientFutureBackend()

  private val baseUri: Uri =
    uri"${config.baseUrl.stripSuffix("/")}"

  // ── Builder companion ─────────────────────────────────────────────────────

  object EtherFlowClient:
    def builder(using ec: ExecutionContext): Builder = Builder()

    def create(baseUrl: String)(using ec: ExecutionContext): EtherFlowClient =
      new EtherFlowClient(EtherFlowConfig(baseUrl = baseUrl))

  // ── Public API ────────────────────────────────────────────────────────────

  /** GET — deserialise JSON into T using circe auto-derivation. */
  def get[T: Decoder](path: String, headers: Map[String, String] = Map.empty): Future[T] =
    execute[T](Method.GET, path, body = None, headers)

  /** GET returning a list of T. */
  def getList[T: Decoder](path: String, headers: Map[String, String] = Map.empty): Future[List[T]] =
    execute[List[T]](Method.GET, path, body = None, headers)

  /** POST — serialise body to JSON via circe, deserialise response to T. */
  def post[T: Decoder, B: Encoder](path: String, body: B, headers: Map[String, String] = Map.empty): Future[T] =
    execute[T](Method.POST, path, body = Some(body.asJson.noSpaces), headers)

  /** PUT — serialise body to JSON, deserialise response to T. */
  def put[T: Decoder, B: Encoder](path: String, body: B, headers: Map[String, String] = Map.empty): Future[T] =
    execute[T](Method.PUT, path, body = Some(body.asJson.noSpaces), headers)

  /** PATCH. */
  def patch[T: Decoder, B: Encoder](path: String, body: B, headers: Map[String, String] = Map.empty): Future[T] =
    execute[T](Method.PATCH, path, body = Some(body.asJson.noSpaces), headers)

  /** DELETE. */
  def delete[T: Decoder](path: String, headers: Map[String, String] = Map.empty): Future[T] =
    execute[T](Method.DELETE, path, body = None, headers)

  /** Safe GET — returns EtherFlowResult, never fails the Future. */
  def getResult[T: Decoder](path: String): Future[EtherFlowResult[T]] =
    get[T](path).transform {
      case Success(v) => Success(EtherFlowResult.Success(v))
      case Failure(e: EtherFlowError) => Success(EtherFlowResult.Failure(e))
      case Failure(e) => Success(EtherFlowResult.Failure(EtherFlowError(e.getMessage, cause = Some(e))))
    }

  /** Health check — calls /health on the configured base URL. */
  def checkHealth(): Future[Map[String, Json]] =
    get[Map[String, Json]]("/health").recover { _ =>
      Map("status" -> Json.fromString("DOWN"), "error" -> Json.fromString("Health check failed"))
    }

  def close(): Unit = backend.close()

  // ── Internal ──────────────────────────────────────────────────────────────

  private def resolveUri(path: String): Uri =
    if path.startsWith("http") then uri"$path"
    else uri"$baseUri$path"

  private def execute[T: Decoder](
    method: Method,
    path: String,
    body: Option[String],
    extra: Map[String, String],
    attempt: Int = 0
  ): Future[T] =
    val uri     = resolveUri(path)
    val headers = (config.defaultHeaders ++ extra ++ Map(
      "Accept"     -> "application/json",
      "User-Agent" -> "EtherFlow-Scala-Client/1.0"
    )).map { case (k, v) => Header(k, v) }.toList

    val baseReq = basicRequest
      .method(method, uri)
      .headers(headers*)
      .response(asStringAlways)

    val req = body.fold(baseReq) { b =>
      baseReq.body(b).contentType(MediaType.ApplicationJson)
    }

    req.send(backend).flatMap { resp =>
      val code = resp.code.code
      if code >= 200 && code < 300 then
        parser.decode[T](resp.body) match
          case Right(v)  => Future.successful(v)
          case Left(err) => Future.failed(EtherFlowError(s"[EtherFlow] Decode error: ${err.getMessage}"))
      else if code >= 400 && code < 500 then
        Future.failed(EtherFlowError(s"[EtherFlow] HTTP $code", statusCode = Some(code), responseBody = Some(resp.body)))
      else
        val err = EtherFlowError(s"[EtherFlow] HTTP $code", statusCode = Some(code))
        if attempt < config.maxRetries then
          println(s"[EtherFlow.Scala] Retry ${attempt + 1}/${config.maxRetries}: ${err.message}")
          val delay = config.retryDelay.toMillis * math.pow(2, attempt).toLong
          after(delay.milliseconds)(execute[T](method, path, body, extra, attempt + 1))
        else
          Future.failed(EtherFlowError(s"[EtherFlow] Max retries exceeded. Last: ${err.message}"))
    }.recoverWith { case e: EtherFlowError => Future.failed(e)
      case e =>
        if attempt < config.maxRetries then
          val delay = config.retryDelay.toMillis * math.pow(2, attempt).toLong
          after(delay.milliseconds)(execute[T](method, path, body, extra, attempt + 1))
        else
          Future.failed(EtherFlowError(s"[EtherFlow] Network error: ${e.getMessage}", cause = Some(e)))
    }

  private def after[T](delay: FiniteDuration)(f: => Future[T]): Future[T] =
    val p = Promise[T]()
    val timer = new java.util.Timer(true)
    timer.schedule(new java.util.TimerTask { def run(): Unit = p.completeWith(f) }, delay.toMillis)
    p.future

// ─────────────────────────────────────────────────────────────────────────────
// Builder
// ─────────────────────────────────────────────────────────────────────────────

final class Builder(using ec: ExecutionContext):
  private var config = EtherFlowConfig()

  def baseUrl(url: String): this.type         = { config = config.copy(baseUrl = url); this }
  def timeout(d: Duration): this.type         = { config = config.copy(timeout = d); this }
  def retry(n: Int): this.type                = { config = config.copy(maxRetries = n); this }
  def retryDelay(d: Duration): this.type      = { config = config.copy(retryDelay = d); this }
  def header(k: String, v: String): this.type = {
    config = config.copy(defaultHeaders = config.defaultHeaders + (k -> v)); this
  }
  def build: EtherFlowClient = new EtherFlowClient(config)

// ─────────────────────────────────────────────────────────────────────────────
// Example Usage
// ─────────────────────────────────────────────────────────────────────────────

/*
// build.sbt:
// libraryDependencies ++= Seq(
//   "com.softwaremill.sttp.client4" %% "core"         % "4.x",
//   "com.softwaremill.sttp.client4" %% "circe"        % "4.x",
//   "io.circe"                      %% "circe-generic" % "0.14.x"
// )

import scala.concurrent.ExecutionContext.Implicits.global
import io.circe.generic.auto.*

case class User(id: Option[Int], name: String, email: String) derives Encoder, Decoder

@main def run(): Unit =
  given ec: ExecutionContext = ExecutionContext.global
  val client = EtherFlowClient.builder
    .baseUrl("https://jsonplaceholder.typicode.com")
    .retry(3)
    .build

  // 2. GET
  val userFuture = client.get[User]("/users/1")
  userFuture.foreach(u => println(s"User: ${u.name} — ${u.email}"))

  // 3. GET list
  val users = client.getList[User]("/users")
  users.foreach(us => println(s"Total users: ${us.size}"))

  // 4. POST
  val newUser = User(None, "Alice", "alice@example.com")
  val created = client.post[User, User]("/users", newUser)
  created.foreach(u => println(s"Created: ${u.name}"))

  // 5. Safe result
  client.getResult[User]("/users/999").foreach {
    case EtherFlowResult.Success(u)  => println(s"Found: ${u.name}")
    case EtherFlowResult.Failure(e)  => println(s"Error: ${e.message}")
  }

  Thread.sleep(5000) // wait for async
  client.close()
*/

File: dart-servers/etherflow_client.dart
/// etherflow_client.dart
/// EtherFlow HTTP Client for Dart — standalone (not Flutter)
///
/// A strongly-typed, async/await HTTP client that mirrors the EtherFlow
/// Java/Kotlin builder API. Uses the `http` package for transport.
///
/// Usage:
///   final client = EtherFlowClient.builder()
///     .baseUrl('https://api.example.com')
///     .retry(3)
///     .build();
///
///   final user = await client.get<User>('/users/1', decoder: User.fromJson);
///   final created = await client.post<User>('/users', body: {'name': 'Alice'}, decoder: User.fromJson);

library etherflow_client;

import 'dart:async';
import 'dart:convert';
import 'package:http/http.dart' as http;

// ─────────────────────────────────────────────────────────────────────────────
// Error
// ─────────────────────────────────────────────────────────────────────────────

/// Represents an error from the EtherFlow Dart client.
class EtherFlowException implements Exception {
  final String message;
  final int? statusCode;
  final String? responseBody;

  const EtherFlowException(this.message, {this.statusCode, this.responseBody});

  @override
  String toString() => '[EtherFlow] $message${statusCode != null ? " (HTTP $statusCode)" : ""}';
}

// ─────────────────────────────────────────────────────────────────────────────
// Config
// ─────────────────────────────────────────────────────────────────────────────

/// Configuration for the EtherFlow Dart client.
class EtherFlowConfig {
  final String baseUrl;
  final Duration timeout;
  final int maxRetries;
  final Duration retryDelay;
  final Map<String, String> defaultHeaders;

  const EtherFlowConfig({
    this.baseUrl = '',
    this.timeout = const Duration(seconds: 10),
    this.maxRetries = 3,
    this.retryDelay = const Duration(milliseconds: 200),
    this.defaultHeaders = const {},
  });
}

// ─────────────────────────────────────────────────────────────────────────────
// Safe Result
// ─────────────────────────────────────────────────────────────────────────────

/// A discriminated union result that never throws — mirrors EtherFlow Java's `toResult()`.
sealed class EtherFlowResult<T> {}

final class EtherFlowSuccess<T> extends EtherFlowResult<T> {
  final T data;
  const EtherFlowSuccess(this.data);
}

final class EtherFlowFailure<T> extends EtherFlowResult<T> {
  final EtherFlowException error;
  const EtherFlowFailure(this.error);
}

// ─────────────────────────────────────────────────────────────────────────────
// Client
// ─────────────────────────────────────────────────────────────────────────────

/// EtherFlow HTTP client for Dart.
/// Mirrors the Java/Kotlin EtherFlowClient builder API with native Dart idioms.
class EtherFlowClient {
  final EtherFlowConfig _config;
  final http.Client _httpClient;

  EtherFlowClient._(this._config, this._httpClient);

  factory EtherFlowClient(EtherFlowConfig config) =>
      EtherFlowClient._(config, http.Client());

  // ── Factory / Builder ────────────────────────────────────────────────────

  static EtherFlowClientBuilder builder() => EtherFlowClientBuilder._();

  static EtherFlowClient create(String baseUrl) =>
      EtherFlowClient(EtherFlowConfig(baseUrl: baseUrl));

  void close() => _httpClient.close();

  // ── Public API ───────────────────────────────────────────────────────────

  /// GET request — decoder maps the JSON Map to your model class.
  Future<T> get<T>(
    String path, {
    required T Function(Map<String, dynamic>) decoder,
    Map<String, String> headers = const {},
  }) async {
    final json = await _executeJson('GET', path, headers: headers);
    return decoder(json as Map<String, dynamic>);
  }

  /// GET a list of objects.
  Future<List<T>> getList<T>(
    String path, {
    required T Function(Map<String, dynamic>) decoder,
    Map<String, String> headers = const {},
  }) async {
    final json = await _executeJson('GET', path, headers: headers);
    return (json as List).map((e) => decoder(e as Map<String, dynamic>)).toList();
  }

  /// POST with JSON body.
  Future<T> post<T>(
    String path, {
    required T Function(Map<String, dynamic>) decoder,
    Map<String, dynamic> body = const {},
    Map<String, String> headers = const {},
  }) async {
    final json = await _executeJson('POST', path, body: body, headers: headers);
    return decoder(json as Map<String, dynamic>);
  }

  /// PUT with JSON body.
  Future<T> put<T>(
    String path, {
    required T Function(Map<String, dynamic>) decoder,
    Map<String, dynamic> body = const {},
    Map<String, String> headers = const {},
  }) async {
    final json = await _executeJson('PUT', path, body: body, headers: headers);
    return decoder(json as Map<String, dynamic>);
  }

  /// PATCH with JSON body.
  Future<T> patch<T>(
    String path, {
    required T Function(Map<String, dynamic>) decoder,
    Map<String, dynamic> body = const {},
    Map<String, String> headers = const {},
  }) async {
    final json = await _executeJson('PATCH', path, body: body, headers: headers);
    return decoder(json as Map<String, dynamic>);
  }

  /// DELETE request.
  Future<T> delete<T>(
    String path, {
    required T Function(Map<String, dynamic>) decoder,
    Map<String, String> headers = const {},
  }) async {
    final json = await _executeJson('DELETE', path, headers: headers);
    return decoder(json as Map<String, dynamic>);
  }

  /// Raw GET — returns decoded JSON as dynamic.
  Future<dynamic> getRaw(String path, {Map<String, String> headers = const {}}) =>
      _executeJson('GET', path, headers: headers);

  /// Safe GET — never throws, returns [EtherFlowResult].
  Future<EtherFlowResult<T>> getResult<T>(
    String path, {
    required T Function(Map<String, dynamic>) decoder,
  }) async {
    try {
      final value = await get<T>(path, decoder: decoder);
      return EtherFlowSuccess(value);
    } on EtherFlowException catch (e) {
      return EtherFlowFailure(e);
    } catch (e) {
      return EtherFlowFailure(EtherFlowException(e.toString()));
    }
  }

  /// Health check — calls `/health` on the base URL.
  Future<Map<String, dynamic>> checkHealth() async {
    try {
      return await getRaw('/health') as Map<String, dynamic>;
    } catch (_) {
      return {'status': 'DOWN', 'error': 'Health check failed'};
    }
  }

  // ── Internal ─────────────────────────────────────────────────────────────

  Uri _buildUri(String path) {
    if (path.startsWith('http')) return Uri.parse(path);
    final base = _config.baseUrl.endsWith('/')
        ? _config.baseUrl.substring(0, _config.baseUrl.length - 1)
        : _config.baseUrl;
    return Uri.parse('$base$path');
  }

  Map<String, String> _buildHeaders(Map<String, String> extra) => {
        'Accept': 'application/json',
        'Content-Type': 'application/json',
        'User-Agent': 'EtherFlow-Dart-Client/1.0',
        ..._config.defaultHeaders,
        ...extra,
      };

  Future<dynamic> _executeJson(
    String method,
    String path, {
    Map<String, dynamic>? body,
    Map<String, String> headers = const {},
  }) async {
    final uri = _buildUri(path);
    final allHeaders = _buildHeaders(headers);
    final bodyBytes = body != null ? utf8.encode(jsonEncode(body)) : null;

    Exception lastError = EtherFlowException('Unknown error');
    var delay = _config.retryDelay;

    for (int attempt = 0; attempt <= _config.maxRetries; attempt++) {
      try {
        final request = http.Request(method, uri);
        request.headers.addAll(allHeaders);
        if (bodyBytes != null) request.bodyBytes = bodyBytes;

        final streamedResponse = await _httpClient
            .send(request)
            .timeout(_config.timeout);
        final response = await http.Response.fromStream(streamedResponse);

        if (response.statusCode >= 200 && response.statusCode < 300) {
          return jsonDecode(response.body);
        }

        final err = EtherFlowException(
          'HTTP ${response.statusCode}: ${response.reasonPhrase}',
          statusCode: response.statusCode,
          responseBody: response.body,
        );

        // Don't retry 4xx
        if (response.statusCode >= 400 && response.statusCode < 500) throw err;
        lastError = err;

      } catch (e) {
        lastError = e is Exception ? e : EtherFlowException(e.toString());
        if (e is EtherFlowException && e.statusCode != null &&
            e.statusCode! >= 400 && e.statusCode! < 500) rethrow;
      }

      if (attempt < _config.maxRetries) {
        print('[EtherFlow.Dart] Retry ${attempt + 1}/${_config.maxRetries} after ${delay.inMilliseconds}ms');
        await Future.delayed(delay);
        delay = delay * 2;
      }
    }
    throw EtherFlowException('[EtherFlow] Max retries exceeded. Last: $lastError');
  }
}

// ─────────────────────────────────────────────────────────────────────────────
// Builder
// ─────────────────────────────────────────────────────────────────────────────

class EtherFlowClientBuilder {
  String _baseUrl = '';
  Duration _timeout = const Duration(seconds: 10);
  int _maxRetries = 3;
  Duration _retryDelay = const Duration(milliseconds: 200);
  final Map<String, String> _headers = {};

  EtherFlowClientBuilder._();

  EtherFlowClientBuilder baseUrl(String url) {
    _baseUrl = url;
    return this;
  }

  EtherFlowClientBuilder timeout(Duration duration) {
    _timeout = duration;
    return this;
  }

  EtherFlowClientBuilder retry(int count) {
    _maxRetries = count;
    return this;
  }

  EtherFlowClientBuilder retryDelay(Duration delay) {
    _retryDelay = delay;
    return this;
  }

  EtherFlowClientBuilder header(String key, String value) {
    _headers[key] = value;
    return this;
  }

  EtherFlowClient build() => EtherFlowClient(EtherFlowConfig(
        baseUrl: _baseUrl,
        timeout: _timeout,
        maxRetries: _maxRetries,
        retryDelay: _retryDelay,
        defaultHeaders: Map.unmodifiable(_headers),
      ));
}

// ─────────────────────────────────────────────────────────────────────────────
// Example Usage
// ─────────────────────────────────────────────────────────────────────────────

/*
 * Run with: dart run etherflow_client.dart

class User {
  final int id;
  final String name;
  final String email;

  const User({required this.id, required this.name, required this.email});

  factory User.fromJson(Map<String, dynamic> json) => User(
        id: json['id'] as int,
        name: json['name'] as String,
        email: json['email'] as String,
      );

  @override
  String toString() => 'User($id, $name, $email)';
}

Future<void> main() async {
  // 1. Create client using builder
  final client = EtherFlowClient.builder()
      .baseUrl('https://jsonplaceholder.typicode.com')
      .retry(3)
      .timeout(const Duration(seconds: 10))
      .build();

  try {
    // 2. GET — auto-decode to User
    final user = await client.get('/users/1', decoder: User.fromJson);
    print('User: ${user.name} — ${user.email}');

    // 3. GET list
    final users = await client.getList('/users', decoder: User.fromJson);
    print('Total users: ${users.length}');

    // 4. POST with body
    final raw = await client.getRaw('/posts');
    print('Posts count: ${(raw as List).length}');

    // 5. Safe result — never throws
    final result = await client.getResult('/users/999', decoder: User.fromJson);
    switch (result) {
      case EtherFlowSuccess(:final data):
        print('Found: ${data.name}');
      case EtherFlowFailure(:final error):
        print('Error: $error');
    }

    // 6. Health check
    final health = await client.checkHealth();
    print('Status: ${health['status']}');

  } finally {
    client.close();
  }
}
*/
